engine/consensus/compliance/core.go

```diff
@@ -137,6 +146,20 @@ func (c *Core) OnBlockProposal(originID flow.Identifier, proposal *messages.Bloc
 		return fmt.Errorf("could not check proposal: %w", err)
 	}
 
+	// ignore proposals which are too far ahead of our local finalized state
+	// instead, rely on sync engine to catch up finalization more effectively, and avoid
+	// large subtree of blocks to be cached.
+	final, err := c.state.Final().Head()
+	if err != nil {
+		return fmt.Errorf("could not get latest finalized header: %w", err)
+	}
```

Comment:
I would suggest using [FinalizedHeaderCache](https://github.com/onflow/flow-go/blob/a2c545d78087b26f97ccea09022f61c3b124d5bb/engine/common/synchronization/finalized_snapshot.go#L32) instead, so we can avoid accessing the state every time.